### PR TITLE
Seed gauge v0.1 review

### PR DESCRIPTION
**Pull request by Brean0**
# Summary
adds/updates the following functionalities
- liquidityWeight: adds a scalar on the effective liquidity that beanstalk measures for an asset.
- germination: introduces a 2 season delay for new deposits eligible for bean mints. Deposits that are in this period are considered "germinating". Germinating deposits do not add to a users stalk and global bdv/amount/stalk.
- Flood v1.1: upon a flood, beanstalk sells beans onto the beanETH well, rather than the bean3CRV metapool. Updated to support a single well implementation. 
- SiloV3.1: The germination update determines whether a deposit is germinating based on its stem. This requires that all silo deposits have a non zero seed value. In Silo V3, stems are truncated by 6 decimals, meaning that the minimum difference between stems are 1 seed worth of grown stalk per BDV. Silo V3.1 increases the stem precision by 6 decimals. This will require a soft migration to update the existing stems of the deposit (no action is required on the user).

# Technicals
## Liquidity Weight
Each silo token now has a "liquidityWeight" parameter when whitelisting. This determines what % of liquidity of the asset beanstalk measures (0 - 100%) during the gauge system. In practice, each token stores a liquidityweight selector that beanstalk calls. 
### AppStorage
- Added `byte4 lwSelector` 
- Decreased `optimalPercentDepositedBdv` to uint64. This is acceptable given this variable has 6 decimal precision, and uint64.max >> 100e6. 
### WhitelistFacet
- Added `lwSelector` parameter to whitelist function 
### Init
- Updated init script to include updating the lwSelector of all gauge tokens.
### SeasonFacet
- Updated LibEvaluate to use the liquidity weight.

## Germination 
Germination has to be handled in the global and farmer level. At a farmer level, only the germinating stalk is tracked. At a global level, the deposited amount and bdv of odd and even germinating is tracked, and the unclaimed germinating assets are stored in a mapping to seasons.
### AppStorage 
- Added `oddGerminating` , `evenGerminating`, and `unlclaimedGermination` to global state
- Added `farmerGerminating` to farmer State
### LibGerminating
- Added LibGerminating for germination logic
### LibSilo
- added `germinating` counterparts to mint/burn
- original mint/burn function are prefixed with `active` to better distinguish between germinating and non germinating actions.
- added `endAccountGermination` that occurs on a mow. 
### LibTokenSilo
- added `germinating` counter parts to incrementing/decrementing deposited bdv/amount
### SeasonFacet
- added `end total germination` to sunrise

## Flood v1.1 
- updated sop to use one singular well pool (bean-eth). Minimal updates are needed to change to another well if desired.

## Silo v3.1 
Silo v3.1 stems now has 6 additional decimal precision. This requires a soft migration.
- in AppStorage, a variable `stemScaleSeason` is added to denote the season in which siloV3.1 was deployed.
- in `LibTokenSilo.removeDepositFromAccount()`, if the input stem % 1e6 == 0, beanstalk checks whether the user has a deposit from the previous silo v3. if they do, it adds that to the total, and `migrates` it to the current silov3.1. This requires the burn/mint events to be emitted to be compliant with ERC1155 in addition to the add/remove deposits. 
- in `libSilo._mow()`, if the user's lastUpdate < stemScaleSeason, update the user's `lastStem` in mowStatuses by 1e6 to correctly match the new precision. Additionally, an edge case must be considered where a user interacts with the silo in the same season as the siloV3.1 update, but prior to the update itself (i.e if the bip was deployed mid-season). This is done by checking the ratio between their lastStems to the stemTip - if stemTip/lastStem >= 1e6. Note that this check fails ~114 years from now, but is acceptable given the small users that this may affect, and that anyone can update a user's silo on their behalf.

### CODE REVIEW

File: protocol/contracts/beanstalk/silo/BDVFacet.sol [hunk 1]
```diff
@@ -18,13 +18,9 @@ import "contracts/libraries/Well/LibWellBdv.sol";
 contract BDVFacet {
     using SafeMath for uint256;
 
-    /**
-     * @dev Returns the BDV of a given `amount` of Bean:3Crv LP tokens.
-     */
     function curveToBDV(uint256 amount) public view returns (uint256) {
         return LibBeanMetaCurve.bdv(amount);
     }
-
```
[the remark this answers is not in the document]
Comment (BrendanSanderson) [reply protocol/contracts/beanstalk/silo/BDVFacet.sol:27]: put the comment back

[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/silo/BDVFacet.sol:27]: https://github.com/BeanstalkFarms/Beanstalk/pull/775/commits/1ecb60a1e2aee572950a158d192d30136ecfad3e

File: protocol/contracts/beanstalk/AppStorage.sol [hunk 2]
```diff
@@ -386,21 +424,26 @@ contract Storage {
      *  uint256 percentOfDepositedBdv
      *  ) external view returns (uint256);
      * ```
-     * @param optimalPercentDepositedBdv The target percentage 
-     * of the total LP deposited BDV for this token.
+     * @param gpSelector The encoded liquidityWeight function selector for the token that pertains to
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/AppStorage.sol:427]: https://github.com/BeanstalkFarms/Beanstalk/pull/775/commits/fee9ee8d15059d50ba8e82cbd199cba765c792c1

File: protocol/contracts/libraries/Silo/LibGerminate.sol [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+        } else if (s.unclaimedGerminating[season.sub(2)].stalk > 0) {
+            s.unclaimedGerminating[season.sub(2)].roots = s
+            .s.roots
+            .mul(s.unclaimedGerminating[season.sub(2)].stalk)
+            .div(s.s.stalk) 
+            .toUint128();
+        }
+        // increment total stalk and roots based on unclaimed values.
+        s.s.stalk = s.s.stalk.add(s.unclaimedGerminating[season.sub(2)].stalk);
+        s.s.roots = s.s.roots.add(s.unclaimedGerminating[season.sub(2)].roots);
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/libraries/Silo/LibGerminate.sol:95]: https://github.com/BeanstalkFarms/Beanstalk/pull/775/commits/12144b72363e3be00669f2e19b443099a4bf7c21

File: protocol/contracts/libraries/Silo/LibGerminate.sol [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+    /**
+     * @notice returns the germinating roots of an account,
+     * as well as clears the germinating stalk and roots.
+     *
+     * @param account address of the account to end germination for.
+     * @param season the season to calculate the germinating roots for.
+     * @param stalk the stalk to calculate the germinating roots for.
+     * @param clearOdd whether to clear the odd or even germinating stalk.
+     */
+    function getGerminatingRoots(
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/libraries/Silo/LibGerminate.sol:187]: [9b0f1d2](https://github.com/BeanstalkFarms/Beanstalk/pull/775/commits/9b0f1d2afcbce0a7eca1c3cc2757b2bda78e4c8c)

File: protocol/contracts/libraries/Silo/LibGerminate.sol [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+                getGerminatingRoots(account, lastMowedSeason.sub(1), secondStalk, !lastUpdateOdd)
+            );
+        }
+
+        // increment users stalk and roots.
+        s.a[account].s.stalk = s.a[account].s.stalk.add(germinatingStalk);
+        s.a[account].roots = s.a[account].roots.add(roots);
+
+        // emit event.
+        emit LibSilo.StalkBalanceChanged(account, int256(germinatingStalk), int256(roots));
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/libraries/Silo/LibGerminate.sol]: https://github.com/BeanstalkFarms/Beanstalk/pull/775/commits/a8eff77ab97f0d7445536d2bc18b262598bd1757

File: protocol/contracts/beanstalk/silo/SiloFacet/SiloGettersFacet.sol [hunk 6]
```diff
@@ -91,11 +215,33 @@ contract SiloExit is ReentrancyGuard {
...
+     * @dev Germinating Stalk that will finish germination upon a silo interaction
+     * is not included.
+     */
+    function balanceOfGerminatingStalk(address account) external view returns (uint256) {
+        (uint256 germinatingStalk, ) = LibGerminate.getCurrentGerminatingStalkAndRoots(
+            account,
+            s.a[account].lastUpdate
+        );
+        return germinatingStalk;
     }
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/silo/SiloFacet/SiloGettersFacet.sol:283]: [011a338](https://github.com/BeanstalkFarms/Beanstalk/pull/775/commits/011a3386ba9e3b905fd388ebcea83b96d5c65fd7)

File: protocol/contracts/beanstalk/silo/SiloFacet/SiloGettersFacet.sol [hunk 7]
```diff
@@ -50,12 +45,118 @@ contract SiloExit is ReentrancyGuard {
...
+    function getGerminatingTotalDeposited(address token) external view returns (uint256 amount) {
+        ( , amount) = LibGerminate.getTotalGerminatingForToken(token);
+    }
+
+    /**
+     * @notice returns the germinating deposited bdv of `token` for the current season.
+     */
+    function getGerminatingTotalDepositedBdv(address token) external view returns (uint256 _bdv) {
+        (_bdv, ) = LibGerminate.getTotalGerminatingForToken(token);
+    }
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/silo/SiloFacet/SiloGettersFacet.sol:94]: added in germ emit pr

File: protocol/contracts/beanstalk/silo/SiloFacet/SiloGettersFacet.sol [hunk 8]
```diff
@@ -64,24 +165,47 @@ contract SiloExit is ReentrancyGuard {
...
         return s.s.roots;
     }
 
+    /**
+     * @notice Returns the total germinating roots for a season.
+     */
+    function getGerminatingRootsForSeason(uint32 season) external view returns (uint256) {
+        return s.unclaimedGerminating[season].roots;
+    }
+
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/silo/SiloFacet/SiloGettersFacet.sol:239]: https://github.com/BeanstalkFarms/Beanstalk/pull/775/commits/75ddb8a55d942dd2f5c0f47b8c559a501e9deeca

File: protocol/contracts/beanstalk/silo/SiloFacet/SiloGettersFacet.sol [hunk 9]
```diff
@@ -353,28 +470,17 @@ contract SiloExit is ReentrancyGuard {
     /**
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/silo/SiloFacet/SiloGettersFacet.sol:508]: https://github.com/BeanstalkFarms/Beanstalk/pull/775/commits/a901a5fd32bc140c1bcf37142beb83a142f7a738

File: protocol/contracts/beanstalk/silo/SiloFacet/SiloGettersFacet.sol [hunk 10]
```diff
@@ -64,24 +165,47 @@ contract SiloExit is ReentrancyGuard {
...
+     * @notice Returns the unclaimed germinating stalk for a season.
+     */
+    function getGerminatingStalkForSeason(uint32 season) external view returns (uint256) {
+        return s.unclaimedGerminating[season].stalk;
+    }
+
+    /**
+     * @notice returns the stalk that is currently in the germination process.
+     */
+    function getTotalGerminatingStalk() external view returns (uint256) {
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/silo/SiloFacet/SiloGettersFacet.sol:182]: https://github.com/BeanstalkFarms/Beanstalk/pull/775/commits/b79419b905d3c27f91ee630e25692529c7a569ca

File: protocol/contracts/libraries/Silo/LibGerminate.sol [hunk 11]
```diff
@@ -0,0 +1,431 @@
...
+            );
+        }
+    }
+
+    /**
+     * @notice returns the stalk and roots currently germinating.
+     * Does not include germinating stalk and roots that will finish germinating
+     * upon an interaction with the silo.
+     */
+    function getCurrentGerminatingStalkAndRoots(
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/libraries/Silo/LibGerminate.sol:292]: https://github.com/BeanstalkFarms/Beanstalk/pull/775/commits/90a7308e30a05179e94c5eff21c471cf4e12ef45

File: protocol/contracts/beanstalk/silo/SiloFacet/TokenSilo.sol [hunk 12]
```diff
@@ -359,131 +409,52 @@ contract TokenSilo is Silo {
...
-        LibSilo.transferStalk(
+        // emit RemoveDeposits event (tokens removed are summation).
+        emit RemoveDeposits(
             sender,
-            recipient,
-            ar.stalkRemoved
+            token,
+            stems, 
+            amounts,
+            ar.active.tokens.add(ar.odd.tokens).add(ar.even.tokens),
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/silo/SiloFacet/TokenSilo.sol:453]: https://github.com/BeanstalkFarms/Beanstalk/pull/775/commits/ea16972b0edc97e5ee2f6e31e157dee2c263bf5d

File: protocol/contracts/libraries/Silo/LibSilo.sol [hunk 13]
```diff
@@ -331,6 +321,66 @@ library LibSilo {
...
+        AssetsRemoved memory ar
+    ) internal {
+        AppStorage storage s = LibAppStorage.diamondStorage();
+        uint256 stalkPerBDV = s.ss[token].stalkIssuedPerBdv;
+        if (ar.active.stalk > 0) {
+            ar.active.stalk = ar.active.stalk.add(ar.active.bdv.mul(stalkPerBDV));
+            transferStalk(sender, recipient, ar.active.stalk);
+        }
+
+        if (ar.odd.stalk > 0) {
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/libraries/Silo/LibSilo.sol:384]: https://github.com/BeanstalkFarms/Beanstalk/pull/775/commits/87d51ed8e2c887166feba3e10570e596a6acd934

File: protocol/contracts/beanstalk/silo/SiloFacet/Silo.sol [hunk 14]
```diff
@@ -148,11 +152,14 @@ contract Silo is SiloExit {
      * with an amount of 0.
      */
     function _claimPlenty(address account) internal {
-        // Plenty is earned in the form of 3Crv.
+        // Plenty is earned in the form of weth.
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/silo/SiloFacet/Silo.sol:155]: https://github.com/BeanstalkFarms/Beanstalk/pull/775/commits/1218fe7ef22649fc4b0773308c9ae1287c13dd7b

File: protocol/contracts/beanstalk/silo/ConvertFacet.sol [hunk 15]
```diff
@@ -113,52 +120,43 @@ contract ConvertFacet is ReentrancyGuard {
...
+
+            // get germinating stem and stemTip for the token
+            LibGerminate.GermStem memory germStem = LibGerminate.getGerminatingStem(token);
+
+            while ((i < stems.length) && (a.active.tokens < maxTokens)) {
+                // skip any stems that are germinating, due to the ability to 
+                // circumvent the germination process.
+                // subtract amount from maxTokens.
+                if (germStem.germinatingStem <= stems[i]) {
+                    maxTokens = maxTokens <= amounts[i] ? 0 : maxTokens.sub(amounts[i]);
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/silo/ConvertFacet.sol:137]: https://github.com/BeanstalkFarms/Beanstalk/pull/775/commits/5f4f3b497aa8985a9bec0f2756a4e3ec31b4fdc9

File: protocol/contracts/beanstalk/silo/ConvertFacet.sol [hunk 16]
```diff
@@ -187,47 +185,66 @@ contract ConvertFacet is ReentrancyGuard {
...
         LibTokenSilo.addDepositToAccount(
             msg.sender, 
             token, 
             stem, 
-            amount, 
+            amount,
             bdv,
             LibTokenSilo.Transfer.emitTransferSingle
-        );
+        );
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/silo/ConvertFacet.sol:248]: https://github.com/BeanstalkFarms/Beanstalk/pull/775/commits/b0aa9c4f54b3153d117b0eeb6ad6e96cd3539fef

File: protocol/contracts/beanstalk/sun/GaugePointFacet.sol [hunk 17]
```diff
@@ -17,23 +17,32 @@ contract GaugePointFacet {
...
         uint256 optimalPercentDepositedBdv,
         uint256 percentOfDepositedBdv
     ) external pure returns (uint256 newGaugePoints) {
-        if (percentOfDepositedBdv > optimalPercentDepositedBdv) {
+        if (percentOfDepositedBdv > optimalPercentDepositedBdv.mul(10001).div(10000)) {
             // gauge points cannot go below 0.
             if (currentGaugePoints <= ONE_POINT) return 0;
             newGaugePoints = currentGaugePoints.sub(ONE_POINT);
-        } else {
+        } else if (percentOfDepositedBdv < optimalPercentDepositedBdv.mul(9999).div(10000)) {
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/sun/GaugePointFacet.sol:41]: https://github.com/BeanstalkFarms/Beanstalk/pull/775/commits/1281a4d14799ae559d5c3eadfca283d284e04913

File: protocol/contracts/beanstalk/sun/SeasonFacet/SeasonFacet.sol [hunk 18]
```diff
@@ -46,11 +46,13 @@ contract SeasonFacet is Weather {
...
         require(!s.paused, "Season: Paused.");
         require(seasonTime() > s.season.current, "Season: Still current Season.");
-        stepSeason();
+        uint32 season = stepSeason();
         int256 deltaB = stepOracle();
         uint256 caseId = calcCaseIdandUpdate(deltaB);
+        LibGerminate.endTotalGermination(season, LibWhitelistedTokens.getWhitelistedTokens());
         LibGauge.stepGauge();
         stepSun(deltaB, caseId);
+        // germination should end after beans are issued.
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/sun/SeasonFacet/SeasonFacet.sol:55]: https://github.com/BeanstalkFarms/Beanstalk/pull/775/commits/0df04a2e672f8549fe6e62118c414d7708d2a661

File: protocol/contracts/beanstalk/sun/SeasonFacet/SeasonGettersFacet.sol [hunk 19]
```diff
@@ -186,22 +159,82 @@ contract SeasonGettersFacet {
...
+            .div(LibGauge.BDV_PRECISION);
     }
 
     /**
-     * @notice returns the next season in that beanstalk will update the stalk growth rate.
+     * @notice Gets the stalk per Gauge Point. Used In gauge system.
      */
-    function getNextStalkGrowthRateUpdate() external view returns (uint256) {
-        return uint256(s.seedGauge.lastStalkGrowthUpdate).add(ONE_WEEK);
+    function getStalkPerGp() external view returns (uint256) {
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/sun/SeasonFacet/SeasonGettersFacet.sol:224]: https://github.com/BeanstalkFarms/Beanstalk/pull/775/commits/f4a724f96f0f75bd724f0b67ba8bc4f9c89529c8

test fixed in later commit

File: protocol/contracts/beanstalk/sun/SeasonFacet/SeasonGettersFacet.sol [hunk 20]
```diff
@@ -246,13 +272,28 @@ contract SeasonGettersFacet {
...
      */
     function getTotalUsdLiquidity() external view returns (uint256) {
-        return getBean3CRVLiquidity().add(getBeanEthTwaUsdLiquidity());
+        return getBeanEthTwaUsdLiquidity();
+    }
+
+    /**
+     * @notice returns the weighted beanEth liquidity used in Gauge calculations.
+     */
+    function getWeightedBeanEthTwaUsdLiquidity() public view returns (uint256) {
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/sun/SeasonFacet/SeasonGettersFacet.sol:286]: https://github.com/BeanstalkFarms/Beanstalk/pull/775/commits/f4a724f96f0f75bd724f0b67ba8bc4f9c89529c8

[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/sun/SeasonFacet/SeasonGettersFacet.sol:286]: 3edc346910f9250c23c639a0c6e63ccb3aed2be7

File: protocol/contracts/beanstalk/silo/EnrootFacet.sol [hunk 21]
```diff
@@ -123,56 +140,118 @@ contract EnrootFacet is ReentrancyGuard {
...
+        LibTokenSilo.incrementTotalDepositedBdv(
+            token, 
+            enrootData.bdvAdded.sub(
+                ar.active.bdv
+                    .add(ar.even.bdv)
+                    .add(ar.odd.bdv)
+            )
+        );
+        LibSilo.mintActiveStalk(
+        msg.sender,
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/silo/EnrootFacet.sol:190]: https://github.com/BeanstalkFarms/Beanstalk/pull/775/commits/424cf5804d99d5ac95161db5d9737892d0853f73

File: protocol/contracts/beanstalk/sun/SeasonFacet/Weather.sol [hunk 22]
```diff
@@ -183,15 +198,22 @@ contract Weather is Sun {
...
+            type(uint256).max
+        );
         rewardSop(amountOut);
-        emit SeasonOfPlenty(s.season.current, amountOut, newHarvestable);
+        emit SeasonOfPlenty(s.season.current, sopWell, address(sopToken), amountOut, newHarvestable);
     }
 
     /**
-     * @dev Allocate 3CRV during a Season of Plenty.
+     * @dev Allocate WETH during a Season of Plenty.
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/sun/SeasonFacet/Weather.sol:216]: https://github.com/BeanstalkFarms/Beanstalk/pull/775/commits/034d46c853ad314149e2c6087931ea2d80fb77c7

File: protocol/contracts/libraries/LibEvaluate.sol [hunk 23]
```diff
@@ -283,15 +303,38 @@ library LibEvaluate {
...
+    function getLiquidityWeight(
+        bytes4 lwSelector
+    ) internal view returns (uint256 liquidityWeight) {
+        bytes memory callData = abi.encodeWithSelector(lwSelector);
+        (bool success, bytes memory data) = address(this).staticcall(callData);
+        if (!success) {
+            return 0;
+        }
+        assembly {
+            liquidityWeight := mload(add(data, add(0x20, 0)))
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/libraries/LibEvaluate.sol:337]: no

File: protocol/contracts/beanstalk/sun/SeasonFacet/Weather.sol [hunk 24]
```diff
@@ -160,16 +171,20 @@ contract Weather is Sun {
      * @dev Flood was previously called a "Season of Plenty" (SOP for short).
      * When Beanstalk has been Oversaturated for a Season, Beanstalk returns the
      * Bean price to its peg by minting additional Beans and selling them directly
-     * on Curve. Proceeds  from the sale in the form of 3CRV are distributed to
+     * on the BEANETH well. Proceeds from the sale in the form of WETH are distributed to
```
[the remark this answers is not in the document]
Comment (Brean0) [reply protocol/contracts/beanstalk/sun/SeasonFacet/Weather.sol:174]: https://github.com/BeanstalkFarms/Beanstalk/pull/775/commits/9cb2f5ca5e3facbf41b298114341e4167706c36b

**Pull request merged**